### ISSUE TITLE
how to handle cancelation of tp.system.prompt() ?

### ISSUE THREAD

**Issue by luckman212** [SilentVoid13/Templater#125]
@SilentVoid13 Thank you for adding `tp.system.prompt()` ! 🙏 
This opens up lots of great possibilities.

One thing I realized that's useful is adding `await` to it so you can reference the input later on, e.g.:

example template:
```md
---
<%* var p = await tp.system.prompt('Note Topic?', 'Mischievous cats');
%>date: <% tp.date.now() %>
topic: <%* tR += p ? p : 'blank' %>
---

- Garfield
- Sylvester
- Cheshire Cat
- Heathcliff
- <% tp.file.cursor(1) %>

```

I noticed that if user presses <kbd>ESC</kbd> when modal is open, a blank "Untitled" note gets created anyway- not sure if this can be avoided at all?

**Comment by SilentVoid13** [comment]
Hey @luckman212, thanks for the feedback.
2 solutions are available when it comes to cancelling `tp.system.prompt` (current behavior where it outputs nothing is actually a bug):
- Continue execution and return something blank for the prompt
- Stop execution and throws an error. Something like "Error cancelled prompt". Other commands won't be resolved.

For the newly created file I'm not sure what's the best solution. I actually create the file **before** I parse and resolve the commands in the template, because some internal functions / variables are requiring a `TFile` object (`tp.file.title` for example).

Should I delete the created file if an error happens during the template parsing ? That may be the best solution, tell me what you think.

**Comment by luckman212** [comment]
@SilentVoid13 Hmm. I don't know if there's an easy answer.

Can `tp.system.prompt()` return `null` if the user cancels? That could be checked for in a script and behavior adjusted if it's truly a cancel vs. just submitting an empty `''` string.

I _do_ think that if user cancels or there's an error, then execution should **stop**.

One idea is - if the created file is EMPTY, then delete it. That way, if some content was already generated before the error, it would be preserved. But, if the system prompt was the first item in the template, and the user canceled, no real value in an empty file being created, so that should be safe to delete. (If this is possible...)

**Comment by SilentVoid13** [comment]
I actually first process every command in the templates and **after** adds the full processed content to the file.
One disadvantage of the Eta templating engine is that if one template fails, everything fails (can't get some partial content).

Taking that into account, do you think it's best to delete the file if command parsing fails ?

What I could do is add an option to `tp.system.prompt` named `throw_on_cancel` to throw an error if cancelled.

I think deleting the file if an error occurs during parsing is the best solution, because it doesn't make much sense to have a file with no replaced commands in it.

**Comment by SilentVoid13** [comment]
@luckman212 Ok nice, I'll add that. About the `throw_on_cancel` option, do you think it would be valuable? I'm not sure everyone wants the prompt to throw on error if cancelled. Should i set this to on or off by default? Not sure what's the most common use case.

**Comment by luckman212** [comment]
Yes, I think it's a good idea whenever new flags or features are introduced that could break existing scripts, to always make it default to the previous behavior. So off by default.

**Comment by SilentVoid13** [comment]
Okay, this has been implemented in Templater 1.5.5. Thanks for the feedback!

**Comment by luckman212** [comment]
Awesome! Just tried it, works great 🚀 

One last question, when cancelling via <kbd>ESC</kbd> an error is dumped to the console as well as a Notice toast:
![image](https://user-images.githubusercontent.com/1992842/115763516-89e4bb00-a372-11eb-8c84-5cc0971a2c47.png)

Is there a way to _not_ have these appear? It would be nice to separate actual parsing errors from a simple user cancellation.
